File: src/main/kotlin/Day04.kt
@JvmInline
value class SectionAssignment(private val range: IntRange) {
    fun fullyOverlaps(value: SectionAssignment): Boolean =
        range.contains(value.range.first) && range.contains(value.range.last)

    fun overlaps(value: SectionAssignment): Boolean =
        range.contains(value.range.first) || range.contains(value.range.last)
}


object Day04 {

    fun solvePart1(lines: List<String>): Int =
        lines.mapToSectionAssignments().count { it.hasFullyOverlappingAssignments() }

    fun solvePart2(lines: List<String>): Int =
        lines.mapToSectionAssignments().count { it.hasOverlappingAssignments() }

    private fun List<String>.mapToSectionAssignments() =
        this.map { line ->

            val (first, second) = line.split(",")
                .map { stringAssignment ->
                    val (start, end) = stringAssignment.split("-")
                    SectionAssignment(
                        range = (start.toInt()..end.toInt())
                    )
                }

            Pair(first, second)
        }


    private fun Pair<SectionAssignment, SectionAssignment>.hasFullyOverlappingAssignments(): Boolean =
        first.fullyOverlaps(second) || second.fullyOverlaps(first)

    private fun Pair<SectionAssignment, SectionAssignment>.hasOverlappingAssignments(): Boolean =
        first.overlaps(second) || second.overlaps(first)
}

File: src/main/ts/day04.ts
class Range {
    private readonly start: number;
    private readonly end: number;

    constructor(str: string) {
        const [start, end] = str.split("-").map(Number)

        this.start = start;
        this.end = end;
    }

    contains(range: Range) {
        return this.contains_number(range.start) && this.contains_number(range.end)
    }

    contains_number(number: number) {
        return this.start <= number && this.end >= number;
    }
}

export class Day04 {
    solvePart1(lines: string[]): number {
        const overlapping = lines
            .map(line => {
                const [pair1, pair2] = line.split(',')

                return {x: new Range(pair1), y: new Range(pair2)}
            })
            .filter(({x,y}) => x.contains(y) || y.contains(x)
            ).length

        return overlapping
    }

    solvePart2(lines: string[]): number {
        const overlapping = lines.filter(line => {
            const pair = line.split(',')

            const [first, second] = pair[0].split('-').map(s => Number(s))
            const [third, fourth] = pair[1].split('-').map(s => Number(s))

            if (second < third) {
                return false
            }
            if (third > second) {
                return false
            }
            if (first > third && first > fourth) {
                return false
            } else {
                return true
            }

        }).length

        return overlapping
    }
}